### PR TITLE
Merge dev branch into master

### PR DESCRIPTION
**Pull request by DevCas1**

### CODE REVIEW

**Comment by DevCas1** [comment]
@Flysenberg I'm calling Loop initially, so it will actually congratulate those on the day it's started

File: ScrubBot.Core/Implementations/Bot.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        private async Task OnClientReadyAsync()
+        {
+            _commandOperator = Container.Get<CommandOperator>();
+            await _manager.Guilds.AddGuildsAsync(_client.Guilds).ConfigureAwait(false);
+            await _manager.Users.AddUsersAsync(_client.Guilds).ConfigureAwait(false);
+        }
+
+        private async Task OnMessageReceivedAsync(SocketMessage message) => await (_commandOperator ?? (_commandOperator = Container.Get<CommandOperator>())).ExecuteAsync(message).ConfigureAwait(false);
```
Comment (Flysenberg) [inline ScrubBot.Core/Implementations/Bot.cs]:
```suggestion
        private async Task OnMessageReceivedAsync(SocketMessage message)
        {
            await _commandOperator.ExecuteAsync(message).ConfigureAwait(false);
        }
```

File: ScrubBot.Modules/EventModule.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            await Database.Events.AddAsync(newEvent);
+
+            await ReplyAsync(new EmbedBuilder().CreateSuccess($"Event **{eventTitle}** has been successfully created for **{occurenceDateTimeUTC}** with a max of **{maxSubscribers}** subscribers!"));
+        }
+
+        [Command("JoinEvent"), Summary("Join a specific event")]
+        public async Task JoinEvent(string eventTitle)
+        {
+            bool hasEvent = Database.Events.Any(x => x.Title == eventTitle);
+            bool isSameServer = Database.Events.Any(x => x.Guild.Id == Context.Guild.Id);
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
            bool isSameServer = Database.Events.Any(x => x.Guild.Id == Guild.Id);
```

Comment (Flysenberg) [reply ScrubBot.Modules/EventModule.cs]: Module base class already has access to the Guild and User models from the database.

File: ScrubBot.Modules/EventModule.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            string username = Context.User.Username;
+
+            if (!hasEvent || !isSameServer)
+            {
+                await ReplyAsync(string.Empty, false, new EmbedBuilder().CreateError($"Could not find an event with eventTitle **{eventTitle}** for this server!"));
+                return;
+            }
+            
+            Event _event = Database.Events.First(x => x.Title == eventTitle && x.Guild.Id == Context.Guild.Id);
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
            Event _event = Database.Events.First(x => x.Title == eventTitle && x.Guild.Id == Guild.Id);
```

File: ScrubBot.Modules/EventModule.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+            if (_event.Author.Id == Context.User.Id)
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
            if (_event.Author.Id == User.Id)
```

File: ScrubBot.Modules/EventModule.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            {
+                await ReplyAsync(string.Empty, false, new EmbedBuilder().CreateError($"**{username}** cannot subscribe to their own event!").Build());
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
                await ReplyAsync(new EmbedBuilder().CreateError($"**{username}** cannot subscribe to their own event!"));
```

Comment (Flysenberg) [reply ScrubBot.Modules/EventModule.cs]: Module base class also has a ReplyAsync method that takes an EmbedBuilder directly

File: ScrubBot.Modules/EventModule.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                return;
+            }
+
+            if (_event.Subscribers.Any(x => x.Id == Context.User.Id))
+            {
+                await ReplyAsync(string.Empty, false, new EmbedBuilder().CreateError($"**{username}** is already subscribed to **{eventTitle}**!").Build());
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
                await ReplyAsync(new EmbedBuilder().CreateError($"**{username}** is already subscribed to **{eventTitle}**!"));
```

File: ScrubBot.Modules/EventModule.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                return;
+            }
+
+            if (_event.Subscribers.Count == _event.MaxSubscribers)
+            {
+                await ReplyAsync(string.Empty, false, new EmbedBuilder().CreateError($"Event **{eventTitle}** is already full!"));
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
                await ReplyAsync(new EmbedBuilder().CreateError($"Event **{eventTitle}** is already full!"));
```

File: ScrubBot.Modules/EventModule.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                return;
+            }
+
+            _event.Subscribers.Add(User);
+            await ReplyAsync(string.Empty, false, new EmbedBuilder().CreateSuccess($"**{username}** has successfully joined event **{eventTitle}** ({_event.Subscribers.Count}/{_event.MaxSubscribers})"));
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
            await ReplyAsync(new EmbedBuilder().CreateSuccess($"**{username}** has successfully joined event **{eventTitle}** ({_event.Subscribers.Count}/{_event.MaxSubscribers})"));
```

File: ScrubBot.Modules/EventModule.cs [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        [Command("DeleteEvent"), Summary("Delete one of your events")]
+        public async Task DeleteEvent(string eventTitle)
+        {
+            Event _event = Database.Events.FirstOrDefault(x => x.Guild.Id == Context.Guild.Id && x.Title == eventTitle);
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
            Event _event = Database.Events.FirstOrDefault(x => x.Guild.Id == Guild.Id && x.Title == eventTitle);
```

File: ScrubBot.Modules/EventModule.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            }
+
+            await ReplyAsync(new EmbedBuilder().CreateEventEmbed("Upcoming events", events));
+        }
+
+        // @ScrubBot CreateEvent "Title goes here" "Description goes here" 24/10/2018 20
+        [Command("CreateEvent"), Summary("Create a new event")]
+        public async Task CreateEvent(string eventTitle, string description, DateTime occurenceDateTimeUTC, int maxSubscribers)
+        {
+            if (Database.Events.Any(x => x.Title == eventTitle && x.Guild.Id == Context.Guild.Id))
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
            if (Database.Events.Any(x => x.Title == eventTitle && x.Guild.Id == Guild.Id))
```

File: ScrubBot.Modules/EventModule.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        public EventModule()
+        {
+
+        }
+
+        [Command("ShowEvents"), Summary("Show the first 10 upcoming events")]
+        public async Task ShowEvents(int amount = 10)
+        {
+            List<Event> events = Database.Events
+                .Where(x => x.Guild.Id == Context.Guild.Id)
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
                .Where(x => x.Guild.Id == Guild.Id)
```

File: ScrubBot.Modules/BirthdayModule.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+            {
+                await ReplyAsync(string.Empty,
+                                 false,
+                                 new EmbedBuilder().CreateError($"Sadly, I know no one with a birthday in {birthdayMonth}...\nEither no one in this server has his/her birthday that month, or they forgot to tell me!"));
+                return;
+            }
+
+            string birthdayList = birthdayBois.Aggregate(string.Empty, (current, birthdayBoi) => current + $"{birthdayBoi.Username} ({birthdayBoi.Birthday:dddd, dd MMMM yyyy})\n");
+
+            await ReplyAsync(string.Empty, false, new EmbedBuilder().CreateMessage($"Birthdays in {birthdayMonth}", birthdayList));
```
Comment (Flysenberg) [inline ScrubBot.Modules/BirthdayModule.cs]:
```suggestion
            await ReplyAsync(new EmbedBuilder().CreateMessage($"Birthdays in {birthdayMonth}", birthdayList));
```

File: ScrubBot.Modules/AdminModule.cs [hunk 13]
```diff
@@ -57,5 +59,27 @@ public async Task SetAuditChannel(SocketTextChannel newChannel)
...
+            Event _event = Database.Events.FirstOrDefault();
+
+            var author = _client.GetUser(_event.Author.Id);
+
+            if (author is null)
+            {
+                await ReplyAsync("Event author is null for some f-ing reason");
+            }
+            else
+                await ReplyAsync(string.Empty, false, new EmbedBuilder().CreateSuccess($"Title\t{_event.Title}\n" +
```
Comment (Flysenberg) [inline ScrubBot.Modules/AdminModule.cs]:
```suggestion
                await ReplyAsync(new EmbedBuilder().CreateSuccess($"Title\t{_event.Title}\n" +
```

File: ScrubBot.Modules/AdminModule.cs [hunk 14]
```diff
@@ -13,9 +13,11 @@ namespace ScrubBot.Modules
     [RequireUserPermission(GuildPermission.Administrator, Group = nameof(AdminModule)), RequireOwner(Group = nameof(AdminModule))]
     public class AdminModule : Module
     {
-        public AdminModule()
-        {
+        DiscordSocketClient _client;
 
+        public AdminModule(DiscordSocketClient client)
```
Comment (Flysenberg) [inline ScrubBot.Modules/AdminModule.cs]: Client can be accessed with Context.Client
```suggestion
        public AdminModule()
```

File: ScrubBot.Modules/AdminModule.cs [hunk 15]
```diff
@@ -13,9 +13,11 @@ namespace ScrubBot.Modules
     [RequireUserPermission(GuildPermission.Administrator, Group = nameof(AdminModule)), RequireOwner(Group = nameof(AdminModule))]
     public class AdminModule : Module
     {
-        public AdminModule()
-        {
+        DiscordSocketClient _client;
```
Comment (Flysenberg) [inline ScrubBot.Modules/AdminModule.cs]: Innecessary
```suggestion
```

File: ScrubBot.Modules/AdminModule.cs [hunk 16]
```diff
@@ -13,9 +13,11 @@ namespace ScrubBot.Modules
...
 
+        public AdminModule(DiscordSocketClient client)
+        {
+            _client = client;
```
Comment (Flysenberg) [inline ScrubBot.Modules/AdminModule.cs]:
```suggestion
```

File: ScrubBot.Modules/AdminModule.cs [hunk 17]
```diff
@@ -57,5 +59,27 @@ public async Task SetAuditChannel(SocketTextChannel newChannel)
 
             await ReplyAsync(new EmbedBuilder().CreateSuccess($"Set this server's audit channel to {newChannel.Mention}"));
         }
+
+        [Command("test")]
+        public async Task TestAsync()
+        {
+            Event _event = Database.Events.FirstOrDefault();
+
+            var author = _client.GetUser(_event.Author.Id);
```
Comment (Flysenberg) [inline ScrubBot.Modules/AdminModule.cs]:
```suggestion
            var author = Context.Client.GetUser(_event.Author.Id);
```

File: ScrubBot.Modules/EventModule.cs [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+        [Command("JoinEvent"), Summary("Join a specific event")]
+        public async Task JoinEvent(string eventTitle)
+        {
+            bool hasEvent = Database.Events.Any(x => x.Title == eventTitle);
+            bool isSameServer = Database.Events.Any(x => x.Guild.Id == Context.Guild.Id);
+            string username = Context.User.Username;
+
+            if (!hasEvent || !isSameServer)
+            {
+                await ReplyAsync(string.Empty, false, new EmbedBuilder().CreateError($"Could not find an event with eventTitle **{eventTitle}** for this server!"));
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
                await ReplyAsync(new EmbedBuilder().CreateError($"Could not find an event with eventTitle **{eventTitle}** for this server!"));
```

File: ScrubBot.Modules/EventModule.cs [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            Event _event = Database.Events.FirstOrDefault(x => x.Guild.Id == Context.Guild.Id && x.Title == eventTitle);
+
+            if (_event is null)
+            {
+                await ReplyAsync(new EmbedBuilder().CreateError($"Unable to find event **{eventTitle}**!"));
+                return;
+            }
+
+            if (_event.Author.Id != Context.User.Id || !Context.Guild.GetUser(Context.User.Id).GuildPermissions.Administrator)
```
Comment (Flysenberg) [inline ScrubBot.Modules/EventModule.cs]:
```suggestion
            if (_event.Author.Id != User.Id || !Context.Guild.GetUser(Context.User.Id).GuildPermissions.Administrator)
```

**Review by Flysenberg — CHANGES_REQUESTED** [review]

**Review by Flysenberg — APPROVED** [review]

**Pull request merged**